### PR TITLE
Add support for JSON Patch

### PR DESCRIPTION
**Pull request by edaena**
## Description
As a developer, I can patch any fhir resource that supports patch using Json Patch according to the FHIR spec using the following call: 
  PATCH [base]/[type]/[id] {?_format=[mime-type]} 

## Related issues
Addresses [issue #].

## Testing
PatchTests.cs

## FHIR Team Checklist
- [ ] **Update the title** of the PR to be succinct and less than 50 characters
- [ ] **Add a milestone** to the PR for the sprint that it is merged (i.e. add S47)
- [ ] Tag the PR with the type of update: **Bug**, **Dependencies**, **Enhancement**, or **New-Feature**
- [ ] Tag the PR with **Azure API for FHIR** if this will release to the managed service
- Review [squash-merge requirements](https://github.com/microsoft/fhir-server/blob/master/SquashMergeRequirements.md)

### Semver Change ([docs](https://github.com/microsoft/fhir-server/blob/master/docs/Versioning.md))
Patch|Skip|Feature|Breaking (reason)

### CODE REVIEW

File: src/Microsoft.Health.Fhir.Core/Resources.resx [hunk 1]
```diff
@@ -1,17 +1,17 @@
 ﻿<?xml version="1.0" encoding="utf-8"?>
 <root>
-  <!--
-    Microsoft ResX Schema
-
+  <!-- 
+    Microsoft ResX Schema 
+
```
Comment (edaena) [inline src/Microsoft.Health.Fhir.Core/Resources.resx:5]: Not sure why these spaces are showing up. Will revert.

File: src/Microsoft.Health.Fhir.Shared.Core/Features/Resources/Patch/PatchResourceHandler.cs [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+            Lazy<IConformanceProvider> conformanceProvider,
+            IResourceWrapperFactory resourceWrapperFactory,
+            ResourceIdProvider resourceIdProvider,
+            IAuthorizationService<DataActions> authorizationService,
+            IModelInfoProvider modelInfoProvider,
+            FhirJsonParser fhirJsonParser,
+            ResourceDeserializer resourceDeserializer)
+            : base(fhirDataStore, conformanceProvider, resourceWrapperFactory, resourceIdProvider, authorizationService)
+        {
+            EnsureArg.IsNotNull(modelInfoProvider, nameof(modelInfoProvider));
```
Comment (brendankowitz) [inline src/Microsoft.Health.Fhir.Shared.Core/Features/Resources/Patch/PatchResourceHandler.cs:42]: nit: looks like we're missing EnsureArg on the other params

File: src/Microsoft.Health.Fhir.Shared.Core/Features/Resources/Patch/PatchResourceHandler.cs [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+            try
+            {
+                FhirJsonNode node = (FhirJsonNode)FhirJsonNode.Parse(currentDoc.RawResource.Data);
+                JObject nodeJson = node.ToJObject();
+                string versionId = (string)nodeJson["meta"]["versionId"];
+                string resourceId = message.ResourceKey.Id;
+
+                message.PatchDocument.ApplyTo(nodeJson);
+
+                FhirJsonNode nodePatch = (FhirJsonNode)FhirJsonNode.Create(nodeJson);
```
Comment (brendankowitz) [inline src/Microsoft.Health.Fhir.Shared.Core/Features/Resources/Patch/PatchResourceHandler.cs:87]: nit: this one can just be
```suggestion
                ISourceNode jsonStatement = FhirJsonNode.Create(nodeJson);
```

File: test/Microsoft.Health.Fhir.Shared.Tests.E2E/Rest/PatchTests.cs [hunk 4]
```diff
@@ -25,11 +28,76 @@ public PatchTests(HttpIntegrationTestFixture fixture)
...
+        }
+
+        [Fact]
+        [Trait(Traits.Priority, Priority.One)]
+        public async Task GivenAServerThatSupportsIt_WhenSubmittingAForbiddenPropertyPatch_ThenAnErrorShouldBeReturned()
+        {
+            var poco = Samples.GetDefaultPatient().ToPoco<Patient>();
+            FhirResponse<Patient> response = await _client.CreateAsync(poco);
+
+            string patchDocument = "[{\"op\":\"replace\",\"path\":\"/versionId\",\"value\":\"100\"}]";
```
Comment (brendankowitz) [inline test/Microsoft.Health.Fhir.Shared.Tests.E2E/Rest/PatchTests.cs:86]: Perhaps convert this one to a [Theory] to test for invalid property paths.
Also https://www.hl7.org/fhir/resource-definitions.html#Meta.lastUpdated should not be patched

File: src/Microsoft.Health.Fhir.Shared.Client/FhirClient.cs [hunk 5]
```diff
@@ -217,6 +217,23 @@ public Task<FhirResponse> HardDeleteAsync<T>(T resource, CancellationToken cance
...
+            {
+                Content = new StringContent(body, Encoding.UTF8, "application/json"),
+            };
+            message.Headers.Accept.Add(_mediaType);
+
+            using HttpResponseMessage response = await HttpClient.SendAsync(message, cancellationToken);
+
+            await EnsureSuccessStatusCodeAsync(response);
+
+            // return new FhirResponse(response);
```
Comment (brendankowitz) [inline src/Microsoft.Health.Fhir.Shared.Client/FhirClient.cs:233]: nit: commented code

File: test/Microsoft.Health.Fhir.Shared.Tests.E2E/Rest/PatchTests.cs [hunk 6]
```diff
@@ -25,11 +28,76 @@ public PatchTests(HttpIntegrationTestFixture fixture)
 
         [Fact]
         [Trait(Traits.Priority, Priority.One)]
-        public async Task GivenAServerThatDoesNotSupportIt_WhenSubmittingAPatch_ThenMethodNotAllowedIsReturned()
+        public async Task GivenAServerThatSupportsIt_WhenSubmittingAPatch_ThenServerShouldPatchCorrectly()
```
Comment (brendankowitz) [inline test/Microsoft.Health.Fhir.Shared.Tests.E2E/Rest/PatchTests.cs:31]: Add `[HttpIntegrationFixtureArgumentSets(DataStore.All, Format.Json)]` to the class to only run for Json

File: src/Microsoft.Health.Fhir.Shared.Core/Features/Resources/Patch/PatchResourceHandler.cs [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+            if (currentDoc.IsHistory)
+            {
+                throw new MethodNotAllowedException(Core.Resources.PatchVersionNotAllowed);
+            }
+
+            try
+            {
+                FhirJsonNode node = (FhirJsonNode)FhirJsonNode.Parse(currentDoc.RawResource.Data);
+                JObject nodeJson = node.ToJObject();
+                string versionId = (string)nodeJson["meta"]["versionId"];
```
[the remark this answers is not in the document]
Comment (edaena) [reply src/Microsoft.Health.Fhir.Shared.Core/Features/Resources/Patch/PatchResourceHandler.cs:82]: @brendankowitz are you suggesting I convert this to Resource, then create a ResourceWrapper to extract the versionId?

File: src/Microsoft.Health.Fhir.Shared.Core/Features/Resources/Patch/PatchResourceHandler.cs [hunk 8]
```diff
@@ -0,0 +1,109 @@
...
+                FhirJsonNode nodePatch = (FhirJsonNode)FhirJsonNode.Create(nodeJson);
+                Resource resourcePatch = nodePatch.ToTypedElement(_modelInfoProvider.StructureDefinitionSummaryProvider).ToPoco<Resource>();
+                string narrativePatch = resourcePatch.Scalar(KnownFhirPaths.ResourceNarrative) as string;
+                string narrativeStatusPatch = (string)nodeJson["text"]["status"];
+
+                if (resource.Id != resourcePatch.Id ||
+                    resource.VersionId != resourcePatch.VersionId ||
+                    resource.Meta.LastUpdated.Value != resourcePatch.Meta.LastUpdated.Value ||
+                    narrative != narrativePatch ||
+                    narrativeStatus != narrativeStatusPatch)
```
Comment (edaena) [inline src/Microsoft.Health.Fhir.Shared.Core/Features/Resources/Patch/PatchResourceHandler.cs:91]: @brendankowitz the validation is getting long here. If there's a better place/approach for this, let me know. Also let me know if there's a better way to access the narrative status for validation.

**Comment by brendankowitz** [comment]
Thanks @edaena, I'm going to pull this into a feature branches. I plan for us to make a few small changes before merging it in :)

**Pull request merged**